File: apps/extension/src/tools/emulate.ts
// `tool.emulate` — apply or clear CDP Emulation-domain overrides
// (viewport metrics, user agent, touch) on a session tab, so agents can
// debug mobile page behaviour. Overrides are per-target: new tabs do
// not inherit them, and `off` restores the tab's real environment.
//
// Device presets are resolved CLI-side; this handler only executes the
// concrete override parameters it receives. Each request is merged
// field by field onto the tab's remembered emulation state and the
// merged state is applied as a whole, so a partial request does not
// reset fields set by an earlier one.

import type { DeviceMetricsOverride, UserAgentOverride } from "@/browser-driver/chromium-cdp";
import { ChromiumCdp } from "@/browser-driver/chromium-cdp";
import type { SessionManager } from "@/session-manager/manager";
import type {
  EmulateOverrides,
  EmulateParams,
  EmulateResult,
  RpcError,
  UserAgentMetadata,
} from "@/transport/types";
import {
  type ChromeTabsApi,
  chromeTabsApi,
  enforceAgentWindow,
  isRpcError,
  lookupSession,
  resolveTargetTab,
} from "./shared";

/**
 * Emulation-domain surface the handler needs. Backed by `ChromiumCdp`
 * in production; tests inject a fake. `trackSessionTab` is optional so
 * test doubles need not implement it.
 */
export interface EmulateCdpRunner {
  setDeviceMetricsOverride(tabId: number, metrics: DeviceMetricsOverride): Promise<void>;
  clearDeviceMetricsOverride(tabId: number): Promise<void>;
  setUserAgentOverride(tabId: number, override: UserAgentOverride): Promise<void>;
  setTouchEmulationEnabled(tabId: number, enabled: boolean, maxTouchPoints?: number): Promise<void>;
  trackSessionTab?(sessionId: string, tabId: number): void;
}

export interface EmulateDeps {
  cdp: EmulateCdpRunner;
  tabsApi: ChromeTabsApi;
  signal?: AbortSignal;
}

function defaultEmulateDeps(): EmulateDeps {
  return {
    cdp: new ChromiumCdp(),
    tabsApi: chromeTabsApi,
  };
}

/**
 * Result note explaining the per-target scope of CDP emulation
 * overrides, echoed on every successful call so agents know to
 * re-apply emulation after opening a new tab.
 */
export const EMULATE_SCOPE_NOTE =
  "emulation overrides are per-tab (CDP target): new tabs do not inherit them — re-run `bsk emulate` on each new tab";

/** Convert the wire (snake_case) UA metadata to CDP's camelCase shape. */
export function toCdpUserAgentMetadata(meta: UserAgentMetadata): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  if (meta.brands !== undefined) out.brands = meta.brands;
  if (meta.full_version !== undefined) out.fullVersion = meta.full_version;
  if (meta.platform !== undefined) out.platform = meta.platform;
  if (meta.platform_version !== undefined) out.platformVersion = meta.platform_version;
  if (meta.architecture !== undefined) out.architecture = meta.architecture;
  if (meta.model !== undefined) out.model = meta.model;
  if (meta.mobile !== undefined) out.mobile = meta.mobile;
  return out;
}

function invalidParams(message: string): RpcError {
  return { code: "invalid_params", message };
}

/**
 * Per-tab record of the emulation state applied so far. New requests
 * are merged onto it field by field (see `mergeEmulateOverrides`) and
 * the merged state is applied as a whole, so e.g. a later
 * `--width/--height` does not silently reset the dpr/mobile of an
 * earlier `--device` preset.
 *
 * The record is best-effort: an entry left behind by a closed tab is
 * simply overwritten by the next emulate call that targets the same tab
 * id, and the whole map is lost when the extension service worker
 * reloads — after a reload the next emulate is equivalent to a full
 * (re)set of just the fields it carries.
 */
const tabEmulationStates = new Map<number, EmulateOverrides>();

/** Test hook: drop every per-tab emulation state record. */
export function resetEmulateStatesForTests(): void {
  tabEmulationStates.clear();
}

/**
 * Merge one validated request onto the tab's stored state: fields
 * present in `overrides` overwrite the stored value, absent fields keep
 * it. Returns a fresh object; `stored` is not mutated.
 */
export function mergeEmulateOverrides(
  stored: EmulateOverrides | undefined,
  overrides: EmulateOverrides,
): EmulateOverrides {
  const merged: EmulateOverrides = { ...stored };
  if (overrides.width !== undefined) merged.width = overrides.width;
  if (overrides.height !== undefined) merged.height = overrides.height;
  if (overrides.device_scale_factor !== undefined) {
    merged.device_scale_factor = overrides.device_scale_factor;
  }
  if (overrides.mobile !== undefined) merged.mobile = overrides.mobile;
  if (overrides.user_agent !== undefined) merged.user_agent = overrides.user_agent;
  if (overrides.accept_language !== undefined) {
    merged.accept_language = overrides.accept_language;
  }
  if (overrides.user_agent_metadata !== undefined) {
    merged.user_agent_metadata = overrides.user_agent_metadata;
  }
  if (overrides.touch !== undefined) merged.touch = overrides.touch;
  if (overrides.max_touch_points !== undefined) {
    merged.max_touch_points = overrides.max_touch_points;
  }
  return merged;
}

/**
 * CDP calls run in sequence without rollback: on a mid-way failure the
 * overrides already applied stay in effect, so the error says how to
 * reset them.
 */
function cdpFailed(err: unknown): RpcError {
  const cause = err instanceof Error ? err.message : String(err);
  return {
    code: "cdp_failed",
    message: `${cause} (overrides applied before the failure were not rolled back — use --off to reset)`,
  };
}

/**
 * Cross-field validation for one overrides set. The CLI performs the
 * same checks; the extension re-validates because the wire format is
 * public.
 */
export function validateOverrides(overrides: EmulateOverrides): RpcError | null {
  const hasWidth = overrides.width !== undefined;
  const hasHeight = overrides.height !== undefined;
  if (hasWidth !== hasHeight) {
    return invalidParams("emulate overrides require width and height together");
  }
  if (hasWidth) {
    if (!Number.isSafeInteger(overrides.width) || (overrides.width as number) <= 0) {
      return invalidParams("width must be a positive integer");
    }
    if (!Number.isSafeInteger(overrides.height) || (overrides.height as number) <= 0) {
      return invalidParams("height must be a positive integer");
    }
  }
  if (
    (overrides.device_scale_factor !== undefined || overrides.mobile !== undefined) &&
    !hasWidth
  ) {
    return invalidParams("device_scale_factor and mobile require width and height");
  }
  if (
    overrides.device_scale_factor !== undefined &&
    (!Number.isFinite(overrides.device_scale_factor) || overrides.device_scale_factor <= 0)
  ) {
    return invalidParams("device_scale_factor must be a positive finite number");
  }
  if (
    (overrides.accept_language !== undefined || overrides.user_agent_metadata !== undefined) &&
    overrides.user_agent === undefined
  ) {
    return invalidParams("accept_language and user_agent_metadata require user_agent");
  }
  if (
    overrides.max_touch_points !== undefined &&
    (!Number.isSafeInteger(overrides.max_touch_points) || overrides.max_touch_points < 1)
  ) {
    return invalidParams("max_touch_points must be a positive integer");
  }
  const anyOverride =
    hasWidth ||
    overrides.user_agent !== undefined ||
    overrides.touch !== undefined ||
    overrides.max_touch_points !== undefined;
  if (!anyOverride) {
    return invalidParams(
      "emulate requires at least one override (viewport, user agent, or touch) or off: true",
    );
  }
  return null;
}

/**
 * Handler for `tool.emulate` (called by the daemon over WS).
 *
 * Merges the requested overrides onto the tab's remembered state and
 * applies the merged state in a fixed order — viewport metrics, then
 * UA, then touch — or clears everything (state included) when `off` is
 * set. Raw CDP failures surface as `cdp_failed` (§4.5).
 */
export async function handleEmulate(
  manager: SessionManager,
  params: EmulateParams,
  deps: EmulateDeps = defaultEmulateDeps(),
): Promise<EmulateResult | RpcError> {
  if (deps.signal?.aborted) return { code: "cancelled", message: "emulate aborted" };
  const ctxOrErr = lookupSession(manager, params, "emulate");
  if (isRpcError(ctxOrErr)) return ctxOrErr;
  const ctx = ctxOrErr;
  const target = await resolveTargetTab(manager, ctx, params?.tab_id, deps.tabsApi);
  if (isRpcError(target)) return target;
  if (deps.signal?.aborted) return { code: "cancelled", message: "emulate aborted" };
  const denied = enforceAgentWindow(ctx, target, "emulate");
  if (denied) return denied;

  if (params.off === true) {
    if (params.overrides !== undefined) {
      return invalidParams("emulate off cannot be combined with overrides");
    }
    try {
      if (deps.signal?.aborted) return { code: "cancelled", message: "emulate aborted" };
      deps.cdp.trackSessionTab?.(ctx.sessionId, target.tabId);
      await deps.cdp.clearDeviceMetricsOverride(target.tabId);
      if (deps.signal?.aborted) {
        return { code: "cancelled", message: "emulate aborted after clearing device metrics" };
      }
      await deps.cdp.setTouchEmulationEnabled(target.tabId, false);
      if (deps.signal?.aborted) {
        return { code: "cancelled", message: "emulate aborted after clearing touch emulation" };
      }
      // An empty userAgent string clears the override (CDP convention).
      await deps.cdp.setUserAgentOverride(target.tabId, { userAgent: "" });
    } catch (err) {
      return cdpFailed(err);
    }
    tabEmulationStates.delete(target.tabId);
    return { tab_id: target.tabId, cleared: true, note: EMULATE_SCOPE_NOTE };
  }

  const overrides = params.overrides;
  if (!overrides) {
    return invalidParams("emulate requires overrides or off: true");
  }
  const invalid = validateOverrides(overrides);
  if (invalid) return invalid;

  // Fields absent from this request keep their previously applied
  // values; the merged state is what gets applied (and echoed back).
  const merged = mergeEmulateOverrides(tabEmulationStates.get(target.tabId), overrides);
  try {
    if (deps.signal?.aborted) return { code: "cancelled", message: "emulate aborted" };
    deps.cdp.trackSessionTab?.(ctx.sessionId, target.tabId);
    if (merged.width !== undefined && merged.height !== undefined) {
      await deps.cdp.setDeviceMetricsOverride(target.tabId, {
        width: merged.width,
        height: merged.height,
        deviceScaleFactor: merged.device_scale_factor ?? 0,
        mobile: merged.mobile ?? false,
      });
      if (deps.signal?.aborted) {
        return { code: "cancelled", message: "emulate aborted after applying device metrics" };
      }
    }
    if (merged.user_agent !== undefined) {
      await deps.cdp.setUserAgentOverride(target.tabId, {
        userAgent: merged.user_agent,
        ...(merged.accept_language !== undefined ? { acceptLanguage: merged.accept_language } : {}),
        ...(merged.user_agent_metadata !== undefined
          ? { userAgentMetadata: toCdpUserAgentMetadata(merged.user_agent_metadata) }
          : {}),
      });
      if (deps.signal?.aborted) {
        return { code: "cancelled", message: "emulate aborted after applying user agent" };
      }
    }
    if (merged.touch !== undefined || merged.max_touch_points !== undefined) {
      await deps.cdp.setTouchEmulationEnabled(
        target.tabId,
        merged.touch ?? true,
        merged.max_touch_points,
      );
    }
  } catch (err) {
    return cdpFailed(err);
  }
  // Record the merged state only once it was fully applied.
  tabEmulationStates.set(target.tabId, merged);
  return {
    tab_id: target.tabId,
    cleared: false,
    applied: merged,
    note: EMULATE_SCOPE_NOTE,
  };
}

File: crates/bsk-protocol/src/tools/emulate.rs
//! Device-emulation tool (`tool.emulate`).
//!
//! Applies CDP Emulation-domain overrides (viewport metrics, user
//! agent, touch) to a single tab so an agent can debug mobile page
//! behaviour. Overrides are per-target: they do not propagate to new
//! tabs, and `off` restores the tab's real environment
//! (`Emulation.clearDeviceMetricsOverride` + touch disabled + UA
//! override cleared).
//!
//! Device presets are resolved CLI-side; the wire only carries the
//! concrete overrides the extension should execute.

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

/// One entry of CDP `Emulation.UserAgentBrandVersion` (UA client hints).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
pub struct UserAgentBrandVersion {
    pub brand: String,
    pub version: String,
}

/// Mirror of CDP `Emulation.UserAgentMetadata`; every field is optional
/// and only sent when present.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
pub struct UserAgentMetadata {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub brands: Option<Vec<UserAgentBrandVersion>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub full_version: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub platform: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub platform_version: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub architecture: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub model: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub mobile: Option<bool>,
}

/// Concrete emulation overrides for one tab. Field presence drives the
/// extension: only the overrides whose fields are set are touched, so a
/// call may change just the UA, just the viewport, or any combination.
#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize, JsonSchema)]
pub struct EmulateOverrides {
    /// Viewport width in CSS pixels. Requires `height`.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub width: Option<u32>,
    /// Viewport height in CSS pixels. Requires `width`.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub height: Option<u32>,
    /// Device scale factor (device pixel ratio). Only meaningful with
    /// `width`/`height`; the extension maps `None` to CDP's `0`
    /// ("use the display's native factor").
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub device_scale_factor: Option<f64>,
    /// Emulate a mobile viewport (meta-viewport handling, overlay
    /// scrollbars). Only meaningful with `width`/`height`.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub mobile: Option<bool>,
    /// UA string for `Emulation.setUserAgentOverride`.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub user_agent: Option<String>,
    /// `Accept-Language` header override. Requires `user_agent`.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub accept_language: Option<String>,
    /// UA client-hints metadata. Requires `user_agent`.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub user_agent_metadata: Option<UserAgentMetadata>,
    /// Enable/disable touch emulation
    /// (`Emulation.setTouchEmulationEnabled`). When omitted while
    /// `max_touch_points` is set, touch is enabled.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub touch: Option<bool>,
    /// Touch points reported while touch emulation is on (CDP defaults
    /// to 1 when omitted).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub max_touch_points: Option<u32>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
pub struct EmulateParams {
    pub session_id: String,
    /// Target tab. Defaults to the Agent Window's active tab.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tab_id: Option<i64>,
    /// Clear every emulation override on the tab and restore its real
    /// environment. Mutually exclusive with `overrides`.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub off: Option<bool>,
    /// Overrides to apply. Required unless `off` is set.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub overrides: Option<EmulateOverrides>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
pub struct EmulateResult {
    pub tab_id: i64,
    /// True when overrides were cleared (`off`); false when applied.
    pub cleared: bool,
    /// Echo of the overrides that were applied. Absent when cleared.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub applied: Option<EmulateOverrides>,
    /// Scope note: emulation overrides are per-tab (CDP target) and are
    /// not inherited by new tabs.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub note: Option<String>,
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn params_omit_optional_fields() {
        let p = EmulateParams {
            session_id: "abcd".into(),
            tab_id: None,
            off: None,
            overrides: None,
        };
        let v = serde_json::to_value(&p).unwrap();
        assert_eq!(v, json!({ "session_id": "abcd" }));
        let round: EmulateParams = serde_json::from_value(v).unwrap();
        assert_eq!(round, p);
    }

    #[test]
    fn overrides_round_trip_with_all_fields() {
        let o = EmulateOverrides {
            width: Some(390),
            height: Some(844),
            device_scale_factor: Some(3.0),
            mobile: Some(true),
            user_agent: Some("Mozilla/5.0 (iPhone…)".into()),
            accept_language: Some("zh-CN".into()),
            user_agent_metadata: Some(UserAgentMetadata {
                brands: Some(vec![UserAgentBrandVersion {
                    brand: "Safari".into(),
                    version: "17".into(),
                }]),
                full_version: None,
                platform: Some("iOS".into()),
                platform_version: None,
                architecture: None,
                model: Some("iPhone".into()),
                mobile: Some(true),
            }),
            touch: Some(true),
            max_touch_points: Some(5),
        };
        let v = serde_json::to_value(&o).unwrap();
        assert_eq!(v["width"], json!(390));
        assert_eq!(v["device_scale_factor"], json!(3.0));
        assert_eq!(
            v["user_agent_metadata"]["brands"][0]["brand"],
            json!("Safari")
        );
        // Absent metadata fields stay absent.
        assert!(v["user_agent_metadata"].get("full_version").is_none());
        let round: EmulateOverrides = serde_json::from_value(v).unwrap();
        assert_eq!(round, o);
    }

    #[test]
    fn overrides_default_is_empty_object() {
        let v = serde_json::to_value(EmulateOverrides::default()).unwrap();
        assert_eq!(v, json!({}));
        let round: EmulateOverrides = serde_json::from_value(json!({})).unwrap();
        assert_eq!(round, EmulateOverrides::default());
    }

    #[test]
    fn off_params_round_trip() {
        let p = EmulateParams {
            session_id: "abcd".into(),
            tab_id: Some(7),
            off: Some(true),
            overrides: None,
        };
        let v = serde_json::to_value(&p).unwrap();
        assert_eq!(v, json!({ "session_id": "abcd", "tab_id": 7, "off": true }));
        let round: EmulateParams = serde_json::from_value(v).unwrap();
        assert_eq!(round, p);
    }

    #[test]
    fn result_round_trips() {
        let r = EmulateResult {
            tab_id: 7,
            cleared: false,
            applied: Some(EmulateOverrides {
                width: Some(390),
                height: Some(844),
                ..EmulateOverrides::default()
            }),
            note: Some("per-tab".into()),
        };
        let v = serde_json::to_value(&r).unwrap();
        let round: EmulateResult = serde_json::from_value(v).unwrap();
        assert_eq!(round, r);
    }

    #[test]
    fn cleared_result_omits_applied() {
        let r = EmulateResult {
            tab_id: 7,
            cleared: true,
            applied: None,
            note: None,
        };
        let v = serde_json::to_value(&r).unwrap();
        assert_eq!(v, json!({ "tab_id": 7, "cleared": true }));
    }
}
